packages/@aws-cdk/aws-iam/lib/grant.ts

```diff
@@ -121,7 +121,19 @@ export class Grant implements cdk.IDependable {
       scope: options.resource,
     });
 
-    if (result.success) { return result; }
+    const compareResult = options.grantee.grantPrincipal.principalAccount
+      ? options.resource.account.compareToString(options.grantee.grantPrincipal.principalAccount)
+      : undefined;
+    // if both accounts are tokens, we assume here they are the same
+    const sameAccount: boolean = compareResult?.equalOrBothUnresolved() ??
+      // if the principal doesn't have an account (for example, a service principal),
```

Comment:
That seems wrong. A service principal almost by definition is in a different account.

Reply:
I mean to refer to the default of "true" here.

Reply:
It's confusing, but it is actually correct (we should _not_ edit the resource policy if a principal that has account `undefined`).

If this was reversed, a lot of tests would fail.

Reply:
Can you point to specifics? 

It seems like:

```ts
bucket.grantRead(new ServicePrincipal('sqs.amazonaws.com')); // Or whichever one
```

Should lead to the statement being added to the resource policy, rather than accepting that it cannot be added to the principal policy.

What is the test that is failing if you change this around, and we sure there's not another problem there?